crates/solver/src/liquidity/zeroex.rs

```diff
@@ -47,20 +50,67 @@ impl ZeroExLiquidity {
             .filter_map(|order| TokenPair::new(order.buy_token, order.sell_token));
         let relevant_pairs = self.base_tokens.relevant_pairs(user_order_pairs);
 
-        let filtered_zeroex_orders = zeroex_orders
+        let order_buckets = self.generate_order_buckets(zeroex_orders, relevant_pairs);
+        let filtered_zeroex_orders = self.get_useful_orders(order_buckets, 5);
+
+        Ok(filtered_zeroex_orders)
+    }
+
+    fn generate_order_buckets(
+        &self,
+        zeroex_orders: impl Iterator<Item = OrderRecord>,
+        relevant_pairs: HashSet<TokenPair>,
+    ) -> OrderBuckets {
+        // divide orders in buckets and choose best orders in those buckets
+        let mut buckets: OrderBuckets = Default::default();
+        zeroex_orders
             .filter(|record| {
                 match TokenPair::new(record.order.taker_token, record.order.maker_token) {
                     Some(pair) => relevant_pairs.contains(&pair),
                     None => false,
                 }
             })
-            .filter_map(|record| self.record_into_liquidity(record))
-            .collect();
+            .for_each(|order| {
+                let bucket = buckets
+                    .entry((order.order.maker_token, order.order.taker_token))
+                    .or_insert(vec![]);
+                bucket.push(order);
+            });
+        buckets
+    }
 
-        Ok(filtered_zeroex_orders)
+    /// Get the `orders_per_type` best priced and biggest volume orders.
+    fn get_useful_orders(
+        &self,
+        order_buckets: OrderBuckets,
+        orders_per_type: usize,
+    ) -> Vec<Liquidity> {
+        let mut filtered_zeroex_orders = vec![];
+        order_buckets.into_values().for_each(|mut orders| {
+            if orders.len() <= 2 * orders_per_type {
+                filtered_zeroex_orders.extend(orders);
+                return;
+            }
+            // Sorting to have best priced orders at the end of the vector
+            orders.sort_by(|order_1, order_2| {
+                let price_1 = order_1.order.taker_amount as f64 / order_1.order.maker_amount as f64;
+                let price_2 = order_2.order.taker_amount as f64 / order_2.order.maker_amount as f64;
+                price_2.partial_cmp(&price_1).unwrap()
+            });
+            filtered_zeroex_orders.extend(orders.drain(orders.len() - orders_per_type..));
+
+            orders.sort_by_key(|order| order.metadata.remaining_fillable_taker_amount);
+            orders.reverse();
+            filtered_zeroex_orders.extend(orders.iter().take(orders_per_type).cloned());
```

Comment:
I remembered that instead of moving elements in the `Vec` around one can also just reverse the iterator (`iter().rev()`).
And instead of cloning the orders one can move them by using `into_iter()`.
Together that would look like this:
```rust
filtered_zeroex_orders.extend(orders.into_iter().rev().take(orders_per_type));
```

Reply:
`orders.reverse()` is no longer necessary if you are iterating backwards over the `Vec` with `into_iter().rev()`.
Otherwise you would reverse twice effectively keeping the original order.